test/OpenTelemetry.Tests/Logs/LogRecordTest.cs

```diff
@@ -915,6 +948,33 @@ private class CustomState
         {
             public string Property { get; set; }
         }
+
+        private class ScopeProcessor : BaseProcessor<LogRecord>
+        {
+            private readonly bool buffer;
+
+            public ScopeProcessor(bool buffer)
+            {
+                this.buffer = buffer;
+            }
+
+            public List<object> Scopes { get; } = new();
+
+            public override void OnEnd(LogRecord data)
+            {
+                data.ForEachScope<object>(
+                    (scope, state) =>
+                    {
+                        this.Scopes.Add(scope.Scope);
+                    },
+                    null);
+
+                if (this.buffer)
+                {
+                    data.Buffer();
```

Comment:
this is an internal API right? Why is this user defined processor calling it?

Reply:
What is happening is a custom processor called before the batch processor which then calls the internal API.

Here is the repro example: https://github.com/mbubnich-bunnings/ForEachScopeProblem/blob/15db9715ebbd96aaff290f551f71c489fde7b40e/ForEachScopeProblem/Program.cs#L35-L36

The test basically just simulates that case.

Reply:
got it.. end users now need not worry about this in their processors, as SDK will ensure its correctly buffered,